pydy/tests/test_joints.py

```diff
@@ -0,0 +1,155 @@
+        assert self.child_joint_point.vel(self.child.frame) == 0
+
+    def test_pinjoint_parameters(self):
+        self.pinjoint = PinJoint('pinjoint', self.parent, self.child,
+                                 (1, 0, 0), (0, 1, 0), axis1='x', axis2='y')
+        assert self.pinjoint.axis1 == self.parent.frame.z
+        assert self.pinjoint.axis2 == self.child.frame.y
+        point1 = self.parent.masscenter.locatenew(self.parent.frame, self.parent.frame.z)
+        point2 = self.child.masscenter.locatenew(self.child.frame, self.child.frame.y)
+        assert self.pinjoint.parent_joint_point == point1
+        assert self.pinjoint.child_joint_piont == point2
+        assert self.pinjoint.parent_joint_vector == self.parent.frame.z
+        assert self.pinjoint.child_joint_vector == self.chidl.frame.y
+
+    def test_joint_joint_functions(self):
+
+        # part 1 assining parent-child relationship
+        self.pinjoint.set_parent_child_rel()
+        assert self.child.parent == self.parent
+        assert self.parent.child == self.child
+
+        # part 2 locating joint point in bodies
+        self.pinjoint._locate_joint_point()
+        child_masscenter = self.child.masscenter
+        parent_masscenter = self.parent.masscenter
+        parent_joint_point = parent_masscenter.locatenew(self.name + '_Point', self.parent_joint_vector)
+        child_joint_point = child_masscenter.locatenew(self.name + '_Point', self.child_joint_vector)
+        assert self.pinjoint.parent_joint_vector == parent_joint_point
+        assert self.pinjoint.child_joint_vector == child_joint_point
+
+        # part 3 aligning axis1 and axis2
+        self.join_frames()
+        # sample implementation
+        # self.child.frame.orient(self.parent.frame, 'Axis', [0, self.parent.frame.x])
+        assert cross(self.pinjoint.axis1, self.pinjoint.axis2) == self.child.frame.z
+        assert cross(self.pinjoint.axis2, self.pinjoint.axis1) == - self.parent.frame.z
+        self.align_axes()
+        axis1 = self.pinjoint.axis1
+        axis2 = self.pinjoint.axis2
+        # sample implementation
+        # perpendicular_axis_in_parent = - cross(axis1, axis2)
+        # angle_between_axes = acos(dot(self.pinjoint.axis2, self.pinjoint.axis1))
+        # self.child.frame.orient(self.parent.frame, 'Axis',
+        #                        [acos(dot(axis1, axis2)/(axis1.magnitude() * axis2.magnitude())),
+        #                         cross(axis1, axis2)])
+        assert acos(dot(axis1, axis2)/(axis1.magnitude() * axis2.magnitude())) == 0
+
+        # part 4 adding angular velocity to child w.r.t parent
+        assert self.child.frame.ang_vel_in(self.parent.frame) != 0
+
+    def test_joint_apply_joint(self):
+        # apply_joint() should do everything done above by calling specific funtions
+        self.new_pinjoint = PinJoint('pinjoint', self.parent, self.child)
+        self.new_pinjoint._apply_joint()
```

Comment:
Just to differentiate between functionalities, I create a method apply_joint which is called by init